### ISSUE TITLE
Wobbly Windows effect + KDE-Rounded-Corners results in distortions of the corners

### ISSUE THREAD

**Issue by rdrms** [matinlotfali/KDE-Rounded-Corners#14]
Hi, just wanted to let you know that the current implementation doesn't play nicely with Wobby Windows.

When moving a window, the corners appear and disappear, and from this it is obvious that the corners are some sort of overlay.

I don't know if there's anything that can be done about this without serious reengineering of the effect.

I value consistency, so I have turned off Wobbly Windows because I really like the rounded corners.

Thanks,
rdr

**Comment by matinlotfali** [comment]
Yes. I confirm that this issue exists on all KWin versions.

**Comment by RomanKode** [comment]
Is the rounded corner effect top layer. Probably it could be easier to fork the effects that have animation and have them affect the rounded corner layer.

**Comment by matinlotfali** [comment]
@RomanKode I might know what are you referring to but I never tried it. Let's try it on the pull request #48.

**Comment by matinlotfali** [comment]
I tried #48 with many different numbers but none of them worked.

**Comment by a-parhom** [comment]
It looks like changing numbers here doesn't work, because `KWin::effects->paintWindow(w, mask, region, data);` call is made before the actual window shaping, so all other effects in stack are triggered before that.

**Comment by RomanKode** [comment]
> It looks like changing numbers here doesn't work, because `KWin::effects->paintWindow(w, mask, region, data);` call is made before the actual window shaping, so all other effects in stack are triggered before that.

Would there be a way to expose the outline effect to that function?

**Comment by matinlotfali** [comment]
Now that I am using only one shader to render the whole effect, we might be able to inject the shader into the `KWin::effects->paintWindow(w, mask, region, data);` using `data.shader`. We don't want to replace its own existing shader (actual content) and instead, add my shader on top of it. 

I tried different approaches to push the shader. But I was not successful. I am not certain if it is possible.

**Comment by matinlotfali** [comment]
I will try to see if I can change the code to work like the Invert effect:
https://github.com/KDE/kwin/blob/master/src/effects/invert/invert.cpp

**Comment by matinlotfali** [comment]
I successfully injected the shader into the `KWin::effects->drawWindow` method and all other effects work very nicely with the render, including the Wobbly effect. 

See pull request #55.

The only problem that I couldn't figure out was the **title bar**. I believe my code is identical to the invert effect but I don't know how the invert effect is able to apply its shader on the titlebar but I can't. 

Any Ideas?

![Screenshot_20220726_203137](https://user-images.githubusercontent.com/7337168/181156873-353fe0d1-5c04-423b-85a4-cfbb9b152d24.png)
![Screenshot_20220726_205315](https://user-images.githubusercontent.com/7337168/181157307-44e72449-04d2-4f72-9e63-a1ec02457ee3.png)

**Comment by a-parhom** [comment]
>I successfully injected the shader into the KWin::effects->drawWindow method and all other effects work very nicely with the render, including the Wobbly effect.

Oh, that's great news!

>The only problem that I couldn't figure out was the title bar. I believe my code is identical to the invert effect but I don't know how the invert effect is able to apply its shader on the titlebar but I can't.

Maybe, I'll try to implement similar approach in my fork, so I'll see if something can be done with this.

**Comment by matinlotfali** [comment]
@a-parhom Okay, I got some idea about why it is not working on my title bars. I changed the shader into this code for debugging. 
```
void main(void)
{
    fragColor = vec4(texcoord0.x, 0, texcoord0.y, 1);
}
```
The outcome shows that the title bar is rendering the shader separately and because its height is so small, it doesn't show the outline and its corners. So I guess I need to find a way to let the shader understand that it should treat the title bar differently.

![Screenshot_20220727_082255](https://user-images.githubusercontent.com/7337168/181286813-e32cfc78-b687-4fd8-9ead-a4408eb4ea9a.png)
.
